### PR TITLE
[IMP] stock: revamp inventory menus + reports 

### PR DESCRIPTION
**Pull request by ticodoo**
Many small back2basics changes 
ENT PR: odoo/enterprise#29974
Upgrade PR: https://github.com/odoo/upgrade/pull/3819

---

[IMP] mrp, stock{_account,_picking_batch}: update inventory menus

- reorder/rename/remove menu contents for "Inventory" and "Reporting",
  including:
 - removed "Forecasted Inventory" menuitem since this view is no longer
   considered useful (code for view to be removed in separate commit)
 - removed "Stock Moves" (moves report) menuitem
 - renamed "Product Move" (move lines report) menuitem => "Moves
   History", this will also be reflected in any "History" buttons that
   open this view from other views.
 - made "Inventory Report" menuitem visible only when applicable (i.e.
   multi-location or consignment is active/debug mpde)
 - made "Run Scheduler" menuitem only available in debug mode (
   main_flow_tour updated to skip scheduler click since general flow is
   expected to still the same/work)

Goal of renaming/ordering of menuitems is to clean them up and make them
more intuitive for users.

- "Run Scheduler" in mrp menus has also been made debug only viewable as
  well to mirror the inventory change.

"menu" part of b2b task: 2882539
ENT PR: https://github.com/odoo/enterprise/pull/29974

---

[REM] stock: remove Forecasted Inventory view

The Forecasted Inventory view has become obsolete, therefore we delete
the code associated with this view. Note that the Forecasted Report
still uses the graph view of this view so we leave it. Because the
Forecasted Report still uses this view + stock.warehouse.orderpoint
relies on the report.stock.quantity model, we leave the model as is.

"forecast inventory" part of b2b task: 2882539
ENT PR: https://github.com/odoo/enterprise/pull/29974
Upgrade PR: https://github.com/odoo/upgrade/pull/3819

---

[IMP] stock{_account}: inventory => locations report update

Inventory report has been updated to have:
- better applicability: report renamed to "Locations" and is only
  visible w/Locations, Consignment, or debug mode active
- improved UX: list view rearranged, buttons added, grouping removed,
  sums added to bottom of list
- improved quant "Value": instead of the accounting value, this is now
  the average unit cost (i.e. sum(valuation layer values)/sum(valuation
  layer quantities) per product x on hand qty (of quant).
- single click load for products
- search based on Warehouse option (including storing
  location.warehouse_id to avoid overly complex search function)
- always show "Location" column in this view even if multi-locations is
  not active

"inventory report" part of b2b task: 2882539

Upgrade PR: https://github.com/odoo/upgrade/pull/3819

---

[IMP] stock: update UX of stock moves report

Minor UX updates to stock moves report:

- don't default groupby source location
- order list by date
- rename "Units of Measure" column to "Unit"
- add color to quantity field (green = incoming qty into stock, red =
  outgoing, black = all others [i.e. internal transfers/external
  locations only, etc]
- search filter order updated
- make to location_{dest_}id text-muted when not a physical location
  (i.e. internal/transit)
- display "Quantity" instead of "Demand" for product_uom_qty and show
  sum at bottom of list

"stock moves" part of b2b task: 2882539

---

[IMP] stock: update UX of stock moves lines report

Note this is specific to the view shown when clicking on:
- "History" button in inventory adjustments
- "Product Moves" button in scrap/product forms
At time of this commit being written. Other move line views not changed
since they are specific to inputting data rather than readonly data.

Minor UX updates to stock move lines report:
- rename "Units of Measure" column to "Unit"
- add color to quantity field (green = incoming qty into stock, red =
  outgoing, black = all others [i.e. internal transfers/external
  locations only, etc]
- search filter order updated + src/dest location filters replaced with
  generic filter that checks if either meets condition
- add src/dest package to list view
- add status, lot, and from/to dest to kanban view and replace picking
  with reference so inventory adjustments are visible as well
- order by date desc so most recent ones are first
- always display location src/dest to help understanding of moves

Specific to "History" button in Inventory Adjustments / Reporting >
Locations (i.e. inventory report), moved product_id from domain to
default_search for more user analysis flexibility/ease.

"product moves" part of b2b task: 2882539

---

[IMP] stock_account: add remaing quantity on layer with standard price

The purpose is to easily identify useful layers needed in an audit.
This way we could create a filter and only see layers that still have
quantity to deliver

Supports task: 2882539

---

[IMP] stock{_account}: update UX of valuation report

Note this is specific to the view shown when clicking on
"Reporting > Valuation", but this uses the same 'stock.quantity.history'
wizard and related custom js as "Configuration > Locations > Current Stock"

Minor UX updates to valuation report:
- don't default groupby product
- add filters incoming and outgoing (same logic as stock.move filters)
- replace "Inventory at Date" with "Valuation at Date" where appropriate
  (but leave location current stock view untouched)
- create separate view for the valuation report so remaining qty/value
  are displayed, but all other svl viewing actions (including the
  "Valuation at Date") will remain showing the original qty/value to
  support deeper analysis
- don't sum unit_cost when groupby is applied in list view since the
  value won't make sense
- add "reference" to svl which is = to stock.move.reference

"valuation" part of b2b task: 2882539

---

[IMP] {purchase_, sale_}stock{_account}: improve forecast report UX

Minor UX updates to forecast report:

- add missing space (between # and unit + before |) also removed
  superfluous uoms (only need it at the end now that an equation is
  displayed)
- move On Hand Value to under product link (+ variants if they exist)
  and make it a link to the valuation report of the product(s) in
  question
- add link from Quantity on Hand to the locations (i.e. inventory/quants)
  report for the relevant product(s). Note that product template was
  added to Search in order to support default searching due to
  forecasting report structure
- redo the top right forecast values to be On Hand + Incoming - Outgoing
  = Forecasted so we provide more explicit and transparent values.
  Additionally, these values are made to not display decimal digits if
  all of those digits are 0.  Note that forecasted values in table are
  untouched (i.e. Forecasted Inventory = Forecasted is unchanged and
  Forecasted with Pending still shows in table)

Also did some t-esc => t-out replacement and other small cleanup since
templates were already being edited.

"forecasted report" part of b2b task: 2882539

---

[IMP] purchase,stock: improve misc small UX

Minor UX updates in purchase and stock:

Reworded (in stock):
- Default name location "Partner Locations" => "Partners"
- Default result for "Apply All" wizard "Inventory Adjustment" =>
  "Quantity Updated"
- Individual "Apply" move reference/name updated for individual quants
  (i.e. "Quantity Updated / Confirmed") so that the scheduled date is
  no longer included

Removed (from purchase):
- Duplicated menuitems, i.e.  "Purchase > Reporting > Purchase" showed
  the same thing as Enterprise menuitem and both were being displayed in
  Enterprise version. Also updated the no records found help message
  since old message appeared to be slightly out of date.

ENT PR: https://github.com/odoo/enterprise/pull/29974
"other" part of b2b task: 2882539

---

[IMP] stock{_account}: add (product) stock report

Adds a new reporting view for [storable] product.product to allow for
easier analysis by product (rather than by quant which is location
specific).

This includes:
- adding 3 new fields for products:
  - avg_cost = total svl value / total qty
  - total_value = avg_cost / on hand
  - company_currency_id = self.env.company.currency_id (i.e. the
    currently selected company) for which currency symbol to display in
    the view
  - IMPORTANT NOTE:
    - these variables (and some of the existing ones) do NOT support
      multi-company aggregations due these values being used to calcuate
      future svls which should be company specific. Larger refactoring
      would be needed to ensure correct calc in both cases. Larger issue
      occurs in stock.report which uses these variables in combination
      with variables that do take into account multi-company records.
- A customized `SearchPanel` so that a warehouse context can be applied
  to the view, rather than a filter. (viewable only when more than 1
  warehouse)

"stock report" part of b2b task: 2882539

---

[IMP] web: allow custom row click action

This commit adds a generic `action="action_some_method"` attribute to
the list view to allow for a custom action when clicking on a (record)
row. This feature mirrors the existing Kanban "action" attribute.

Note that in cases where the action method does not return a valid
action then the default action `act_window_close` will be called
instead (same behavior as the kanban view and buttons in general).

Supports "magic link" parts of Task: 2882539

---

[IMP] stock{_account}: add reference opening magic in list view

The move line form view is too technical for the average user so instead we
can to auto-magically open the reference document instead (picking or MO
only for now).

We do the same with stock valuation layers as well, but this feature
will not be available until its custom JS is converted to OWL

"product moves" part of b2b task: 2882539

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by svs-odoo — COMMENTED** [review]
A bunch a little suggestions but nothing blocking

File: addons/web/static/src/views/list/list_arch_parser.js [hunk 1]
```diff
@@ -153,6 +153,11 @@ export class ListArchParser extends XMLParser {
                 treeAttr.noOpen = archParseBoolean(node.getAttribute("no_open") || "");
                 treeAttr.expand = archParseBoolean(xmlDoc.getAttribute("expand") || "");
                 treeAttr.decorations = getDecoration(xmlDoc);
+
+                // custom open action when clicking on record row
+                const action = xmlDoc.getAttribute("action");
+                const type = xmlDoc.getAttribute("type") || "object";
```
Comment (aab-odoo) [inline addons/web/static/src/views/list/list_arch_parser.js:159]: Not sure about the default value.

`<tree action="abc"/>` => `abc` is a method
`<tree action="abc" type="action"/>` => `abc` is an action

Doesn't look natural as the other attribute is named `action`.

Note: in Kanban, there's no default (same API as for `<button>`)

File: addons/web/static/src/views/list/list_controller.js [hunk 2]
```diff
@@ -163,6 +163,24 @@ export class ListController extends Component {
...
+        if (this.archInfo.openAction) {
+            this.actionService.doActionButton({
+                name: this.archInfo.openAction.action,
+                type: this.archInfo.openAction.type,
+                resModel: record.resModel,
+                resId: record.resId,
+                resIds: record.resIds,
+                context: record.context,
+                onClose: async () => {
+                    this.defaultOpenRecord(record); //TODO: FIXME? this is a hack in case non-valid action returned
```
Comment (aab-odoo) [inline addons/web/static/src/views/list/list_controller.js:175]: Why would we want to handle this case? We should reload the record/view instead (when we close the dialog, if the action opens in a dialog)

**Review by aab-odoo — COMMENTED** [review]
[about the `action` attribute in list view] Good job overall. This needs to be tested though.

File: addons/web/static/src/views/list/list_arch_parser.js (hunk 1 above)
Comment (ticodoo) [reply addons/web/static/src/views/list/list_arch_parser.js]: Agreed and changed it to match the default behavior of the Kanban + Button behavior

File: addons/web/static/src/views/list/list_controller.js (hunk 2 above)
Comment (ticodoo) [reply addons/web/static/src/views/list/list_controller.js]: Done. Removed the unnecessary `defaultOpenRecord` as well.

**Comment by ticodoo** [comment]
@aab-odoo Thanks again for your review. I added 2 QUnit tests (one for type='object' and one for type='action') to match the kanban tests. Please let me know if there are any other changes that should be done.

**Comment by aab-odoo** [comment]
> @aab-odoo Thanks again for your review. I added 2 QUnit tests (one for type='object' and one for type='action') to match the kanban tests. Please let me know if there are any other changes that should be done.

Good job :+1:

**Comment by amoyaux** [comment]
robodoo r+ rebase-ff

File: addons/stock/views/stock_quant_views.xml [hunk 3]
```diff
@@ -174,15 +190,13 @@
...
-                search_default_locationgroup=1,
+                always_show_loc=1,
             ).action_view_quants()
         </field>
     </record>
 
     <record model="ir.actions.act_window" id="dashboard_open_quants"> <!-- Used in dashboard -->
-        <field name="name">Stock On Hand</field>
-        <field name="context">{'search_default_internal_loc': 1, 'search_default_productgroup':1, 'search_default_locationgroup':1}</field>
+        <field name="name">Locations</field>
```
Comment (KangOl) [inline addons/stock/views/stock_quant_views.xml:199]: Force an empty value for the context, else the previous one will be kept at the update
```suggestion
        <field name="name">Locations</field>
        <field name="context">{}</field>
```

**Comment by amoyaux** [comment]
robodoo r+

**Comment by amoyaux** [comment]
robodoo r+

**Pull request closed without merging**